File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Boolean}
def is_valid_bst(root)
    def validate(node, min, max)
        return false if node.val<=min || node.val>=max
        if node.left
            return false if !validate( node.left, min, node.val)
        end
        if node.right
             return false if !validate( node.right, node.val, max)
        end
        return true
    end
    min,max=10**31*-1, 10**31
    validate(root, min, max)!=false
end

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
    // if left is null fill the left array with the root and skip checking 
    // if right is null, fill the node with the root and skip checking 
    // check the max node is less thatn the root on the left side 
    // check the min value is greater than the node on the right side 
    let result=true
    function validBST(root){
        
        if(!root||!result)
            return null
        console.log(root.val)
        let left=validBST(root.left)
        let right=validBST(root.right)
        if(result){
            if(left){
                if(left[1]>=root.val){
                    result=false
                    return
                }
            }else{
                left=[root.val, root.val]
            }
            if(right){
                if(right[0]<=root.val){
                    result=false
                    return
                }
            }else{
                right=[root.val, root.val]
            }
            
            return [left[0], right[1]]
        }
        return
    }
    validBST(root)
    return result;
};